src/functional/geometry/convex_hull.zig

```diff
@@ -0,0 +1,243 @@
+        };
+    }
+
+    // Deinitialize the container and free the allocated memory.
+    pub fn deinit(self: *PointSet) void {
+        self.allocator.free(self.points);
+    }
+
+    // Add `point` to the `PointSet`.
+    pub fn add(self: *PointSet, point: Point) !void {
+        if (self.len >= self.max_len) {
+            // Resize the set of `points`.
+            self.max_len = self.max_len * 2;
+            self.points = try self.allocator.realloc(self.points, self.max_len);
+        }
+        self.points[self.len] = point;
+        self.len += 1;
+    }
+
+    // Remove the last point from the set.
+    pub fn pop(self: *PointSet) !void {
+        if (self.len == 0) return Error.EmptySet;
+        self.len -= 1;
+    }
+};
+
+/// Graham algorithm to add a new point to `upper_hull` and `lower_hull`. The algorithm ensures
+/// that the upper and lower hull form a Convex Hull.
+pub fn addPointToConvexHull(upper_hull: *PointSet, lower_hull: *PointSet, point: Point) !void {
+    if (upper_hull.len < 2) {
+        // The first two points can be add directly.
+        try upper_hull.add(point);
+    } else {
+        // Update upper hull.
+        var top: usize = upper_hull.len - 1;
+        while ((top > 0) and (getTurn(
+            upper_hull.points[top - 1],
+            upper_hull.points[top],
+            point,
+        ) != Turn.right)) : (top -= 1) {
+            try upper_hull.pop();
+        }
+        try upper_hull.add(point);
+    }
+
+    if (lower_hull.len < 2) {
+        // The first two points can be add directly.
+        try lower_hull.add(point);
+    } else {
+        // Update lower hull.
+        var top: usize = lower_hull.len - 1;
+        while ((top > 0) and (getTurn(
+            lower_hull.points[top - 1],
+            lower_hull.points[top],
+            point,
+        ) != Turn.left)) : (top -= 1) {
+            try lower_hull.pop();
+        }
+        try lower_hull.add(point);
+    }
+}
+
+/// Return the type of turn created by the `first_point`, `middle_point and the `last_point`.
+fn getTurn(first_point: Point, middle_point: Point, last_point: Point) Turn {
+    const distance_last_middle: f64 = @floatFromInt(last_point.time - middle_point.time);
+    const distance_middle_first: f64 = @floatFromInt(middle_point.time - first_point.time);
+
+    const value = (middle_point.value - first_point.value) * distance_last_middle -
+        (last_point.value - middle_point.value) * distance_middle_first;
+
+    if (value == 0) return Turn.colinear;
+
+    return if (value > 0) Turn.right else Turn.left;
+}
+
+test "incremental convex hull with known result" {
+    const allocator = testing.allocator;
+
+    var upper_hull = try PointSet.init(&allocator, 2);
+    defer upper_hull.deinit();
+    var lower_hull = try PointSet.init(&allocator, 2);
+    defer lower_hull.deinit();
+
+    var point: Point = Point{ .time = 0, .value = 3 };
+    try addPointToConvexHull(&upper_hull, &lower_hull, point);
```

Comment:
Can `point` be replaced with an [Anonymous Struct Literals](https://ziglang.org/documentation/0.12.0/#Anonymous-Struct-Literals), e.g., `try addPointToConvexHull(&upper_hull, &lower_hull, .{.time = 0, .value =3});`? If so, maybe use it to significantly reduce the number of lines in the test.

Reply:
Cool. I followed this advice.